lib/db/observed.go

```diff
@@ -0,0 +1,38 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package db
+
+import (
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+func ListPendingDevices(db *Lowlevel) (map[protocol.DeviceID]ObservedDevice, error) {
```

Comment:
Just `PendingDevices`. Also currently all the usages are on slices - I assume you have an additional usage in mind where a map is required. Otherwise return a slice.

Reply:
The name was inspired by e.g. `func (s *FileSet) ListDevices() []protocol.DeviceID`.  Especially since this is not a real method (no receiver argument), it felt wrong to take a package-global name just consisting of the noun.

My justification for the map was this: it's easy for any potential user to flatten it into a slice (such as the API).  Doing the reverse is needlessly harder, so it mirrors the database structure key -> value at this specific level. Might do the conversion in the model layer, but then again the flat list is actually API-specific.

Reply:
Good point about the non-method, completely missed that: What's the reasoning for that?

We aren't really consistent on the naming, but I think it's just historical: It should generally be just the name for the getter and a `Set` prefix for the setter.

I am not against the map generally. My point is just that if there really never ever is a map access and it only gets flattened, it might as well be a slice from the getgo.

Reply:
> Good point about the non-method, completely missed that: What's the reasoning for that?

It's not implemented in `db/lowlevel.go` so I did not consider it a "method" of `LowLevel` with my C++ / Python / $whatever_OOP$ background.  If it's okay to mix in methods from other modules within the package, we might as well make it so.

> I am not against the map generally. My point is just that if there really never ever is a map access and it only gets flattened, it might as well be a slice from the getgo.

Some code part needs to do it.  The model IMO is a place to access structured data as far as possible.  I don't know yet whether map access will be necessary, but thinking about migration code etc. I suspect the need will arise sooner or later.  Iterating over only keys for example is very easy in Go if you have a map.

Reply:
> We aren't really consistent on the naming, but I think it's just historical: It should generally be just the name for the getter and a `Set` prefix for the setter.

I chose a non-representative example as it seems:

    set.go:456:func DropFolder(db *Lowlevel, folder string) {
    set.go:474:func DropDeltaIndexIDs(db *Lowlevel) {

I understood you prefer to use a method, so working on that.

Reply:
Another argument for using the map was actually my misunderstanding of arrays and slices.  I didn't know that appending to a slice will extend (and reallocate!) the underlying array automatically.  Since I'm iterating through the database entries without knowing their total count in advance, I thought that inserting to a map would be easier / more efficient than extending the slice for each entry.